app/src/device_msg.c

```diff
@@ -56,21 +63,43 @@ sc_device_msg_deserialize(const uint8_t *buf, size_t len,
             if (size < len - 5) {
                 return 0; // not available
             }
-            uint8_t *data = malloc(size);
-            if (!data) {
-                LOG_OOM();
+
+            msg->uhid_output.id = id;
+            msg->uhid_output.size = size;
+            if (read_message(&msg->uhid_output.data, &buf[5], size) == -1) {
                 return -1;
             }
-            if (size) {
-                memcpy(data, &buf[5], size);
+
+            return 5 + size;
+        case DEVICE_MSG_TYPE_MEDIA_UPDATE: {
+            if (len < 5) {
+                // at least id + size
+                return 0; // not available
+            }
+            uint16_t id = sc_read16be(&buf[1]);
+            size_t size = sc_read16be(&buf[3]);
+            if (size < len - 5) {
+                return 0; // not available
             }
 
-            msg->uhid_output.id = id;
-            msg->uhid_output.size = size;
-            msg->uhid_output.data = data;
+            msg->media_update.id = id;
+            msg->media_update.size = size;
+            if (read_message(&msg->media_update.data, &buf[5], size) == -1) {
+                return -1;
+            }
 
             return 5 + size;
         }
+        case DEVICE_MSG_TYPE_MEDIA_REMOVE: {
+            if (len < 3) {
+                // at least id
+                return 0; // not available
+            }
+            uint16_t id = sc_read16be(&buf[1]);
+            msg->media_remove.id = id;
+            return 3;
+            }
+        }
```

Comment:
Is it intentional that these closing brackets don't match their indentation level? The first one closes the case of `DEVICE_MSG_TYPE_MEDIA_REMOVE` (henceforth _m_) while the second one closes the case of `DEVICE_MSG_TYPE_UHID_OUTPUT` (henceforth _o_).

(I'm not even sure if the `case` of _m_ works since it's *inside* the block opened by _o_, but I don't know C well enough to say for sure. Maybe `case`s inside a switch can be in nested blocks and blocks aren't hard scopes.)